unison-cli/src/Unison/Codebase/Editor/HandleInput/FindAndReplace.hs

```diff
@@ -91,6 +93,47 @@ handleStructuredFindI rule = do
   Cli.setNumberedArgs $ map SA.HashQualified results
   Cli.respond (ListStructuredFind results)
 
+handleTextFindI :: Bool -> [String] -> Cli ()
+handleTextFindI allowLib tokens = do
+  Cli.Env {codebase} <- ask
+  currentBranch <- Cli.getCurrentBranch0
+  hqLength <- Cli.runTransaction Codebase.hashLength
+  let names = Branch.toNames currentBranch
+  let ppe = PPED.makePPED (PPE.hqNamer hqLength names) (PPE.suffixifyByHash names)
+  let fqppe = PPED.unsuffixifiedPPE ppe
+  results :: [(HQ.HashQualified Name, Referent)] <- pure $ do
+    r <- Set.toList (Relation.ran $ Names.terms names)
+    Just hq <- [PPE.terms fqppe r]
+    fullName <- [HQ'.toName hq]
+    guard (allowLib || not (Name.beginsWithSegment fullName NameSegment.libSegment))
```

Comment:
Because of how names objects are stored and prefixed this will iterate over every single name from your entire lib dir recursively (though we do have a deduplication by hash).

It's much much faster to use `Branch.withoutLib` or `Branch.withoutTransitiveLibs` on the Branch before converting it to a `Names` object, then we don't do any work on the lib names at all.

Reply:
TIL, thanks!

The structural find and replace in this file can use the same treatment.